src/client.rs

```diff
@@ -280,3 +270,368 @@ where
+                    TransactionProgressStatus::Finalized(TransactionInBlock {
+                        block_hash: hash,
+                        ext_hash: self.ext_hash,
+                        client: self.client,
+                    })
+                }
+                TransactionStatus::Usurped(hash) => {
+                    self.sub = None;
+                    TransactionProgressStatus::Usurped(hash)
+                }
+                TransactionStatus::Dropped => {
+                    self.sub = None;
+                    TransactionProgressStatus::Dropped
+                }
+                TransactionStatus::Invalid => {
+                    self.sub = None;
+                    TransactionProgressStatus::Invalid
+                }
+            }
+        }))
+    }
+
+    /// Wait for the transaction to be in a block (but not necessarily finalized), and return
+    /// an [`TransactionInBlock`] instance when this happens, or an error if there was a problem
+    /// waiting for this to happen.
+    ///
+    /// **Note:** consumes `self`. If you'd like to perform multiple actions as the state of the
+    /// transaction progresses, use [`TransactionProgress::next()`] instead.
+    pub async fn wait_for_in_block(
+        mut self,
+    ) -> Result<TransactionInBlock<'client, T>, Error> {
+        while let Some(status) = self.next().await? {
+            match status {
+                // Finalized or otherwise in a block! Return.
+                TransactionProgressStatus::InBlock(s)
+                | TransactionProgressStatus::Finalized(s) => return Ok(s),
+                // Error scenarios; return the error.
+                TransactionProgressStatus::FinalityTimeout(_) => {
+                    return Err(TransactionError::FinalitySubscriptionTimeout.into())
+                }
+                TransactionProgressStatus::Invalid => {
+                    return Err(TransactionError::Invalid.into())
+                }
+                TransactionProgressStatus::Usurped(_) => {
+                    return Err(TransactionError::Usurped.into())
+                }
+                TransactionProgressStatus::Dropped => {
+                    return Err(TransactionError::Dropped.into())
+                }
+                // Ignore and wait for next status event:
+                _ => continue,
+            }
+        }
+        Err(RpcError::Custom("RPC subscription dropped".into()).into())
+    }
+
+    /// Wait for the transaction to be finalized, and return a [`TransactionInBlock`]
+    /// instance when it is, or an error if there was a problem waiting for finalization.
+    ///
+    /// **Note:** consumes `self`. If you'd like to perform multiple actions as the state of the
+    /// transaction progresses, use [`TransactionProgress::next()`] instead.
+    pub async fn wait_for_finalized(
+        mut self,
+    ) -> Result<TransactionInBlock<'client, T>, Error> {
+        while let Some(status) = self.next().await? {
```

Comment:
Are transport-level errors handled somewhere else or do we ignore them with the blanket `RpcError::Custom` below?

Reply:
Transport level errors are returned by that `?` on the `.await` on this line. The blanket custom error (which I'm not really a fan of, but just copied from the previous approach) is just to handle the case that the subscription is dropped for some reason upstream without telling us any of the expected transaction progress states.

Reply:
fair enough, this is really if the subscription channel stream can't keep up with server.

as we use `mpsc::channel` in jsonrpsee we simply just drop the subscription channel if it gets full, maybe you have ideas to make it more ergonomic? :)

however, this `RpcError::Custom("RPC subscription dropped".into()` "should not during normal circumstances"